File: javascript/day03/part2.js
const { assert } = require('console')
const { getPriority } = require('./utils')
const rucksacks = require('../file-reader.js').readFile('input.txt', '\n', String)

function intersection (a, b, c) {
    const setA = new Set(a);
    const ab = b.filter(value => setA.has(value));
    const setAB = new Set(ab)
    return c.filter(value => setAB.has(value))
}

let prioritySum = 0
for (let i = 0; i < rucksacks.length; i += 3) {
    const group = rucksacks.slice(i, i + 3).map(sack => sack.split(''))
    const badge = intersection(...group)[0]
    const priority = getPriority(badge)
    prioritySum += priority
}

console.log(prioritySum)
assert(prioritySum === 2616, `Incorrect sum of priorities.`)

File: python/day03/part2.py
t = 0

with open('input.txt') as f:
    lines = f.readlines()
    for index in range(0, len(lines), 3):
        x = lines[index].strip()
        y = lines[index+1].strip()
        z = lines[index+2].strip()

        k, = set(x) & set(y) & set(z)
        if k >= "a":
            t += ord(k) - ord("a") + 1
        else:
            t += ord(k) - ord("A") + 27

print(t)
assert t == 2616, 'Incorrect result'
